src/main/java/bio/terra/service/kubernetes/KubePodListener.java

```diff
@@ -43,61 +52,93 @@ public void run() {
 
-                    String operation = item.type;
-                    String podName = item.object.getMetadata().getName();
-                    logger.info(String.format("%s : %s%n", operation, podName));
-
-                    if (StringUtils.contains(podName, API_POD_FILTER)) {
-                        if (StringUtils.equals(operation, "ADDED")) {
-                            logger.info("Added api pod: " + podName);
-                            podMap.put(podName, true);
-                        } else if (StringUtils.equals(operation, "DELETED")) {
-                            // TODO: call Stairway to inform it of a deleted pod
-                            Boolean deletedPod = podMap.get(podName);
-                            if (deletedPod != null && !deletedPod) {
-                                logger.info("Deleted api pod: " + podName);
-                                podMap.put(podName, false);
+                        if (StringUtils.contains(podName, API_POD_FILTER)) {
+                            if (StringUtils.equals(operation, "ADDED")) {
+                                logger.info("Added api pod: " + podName);
+                                podMap.put(podName, true);
+                            } else if (StringUtils.equals(operation, "DELETED")) {
+                                try {
+                                    logger.info("Attempting clean up of deleted stairway instance: " + podName);
+                                    stairway.recoverStairway(podName);
+                                    Boolean deletedPod = podMap.get(podName);
+                                    if (deletedPod != null && !deletedPod) {
+                                        logger.info("Deleted api pod: " + podName);
+                                        podMap.put(podName, false);
+                                    }
+                                } catch (DatabaseOperationException | StairwayExecutionException ex) {
+                                    logger.error("Stairway recoveryStairway failed to recovery pod: " + podName, ex);
+                                } catch (InterruptedException ex) {
+                                    logger.info("KubePodListener interrupted - exiting", ex);
+                                    exception = ex;
+                                    return;
+                                }
                             }
                         }
                     }
                 }
+            } catch (RuntimeException | ApiException | IOException ex) {
+                exception = ex;
+                logger.info("KubePodListener caught exception: " + exception);
             }
-        } catch (ApiException | IOException ex) {
-            exception = ex;
-        }
 
-        logger.info("KubePodListener exiting - exception: ", exception);
+            // Exponential backoff retry after an exception
+            if (consecutiveRetryCount >= WATCH_RETRIES) {
```

Comment:
What kinds of recoverable exceptions were you seeing in starting up the listener?

Reply:
Every now and then, I see the watch quit on a network timeout. This loop will restart it. I didn't want to infinite loop on a hard failure, so I put in the counter for the retries. I have never seen it over 1, though.

Reply:
I added a comment explaining that.

Reply:
I added a comment saying that.